pkg/dockerregistry/server/metrics/errorcodes.go

```diff
@@ -0,0 +1,76 @@
+package metrics
+
+import (
+	"io/fs"
+	"syscall"
+
+	storagedriver "github.com/distribution/distribution/v3/registry/storage/driver"
+)
+
+const (
+	errCodeUnsupportedMethod = "UNSUPPORTED_METHOD"
+	errCodePathNotFound      = "PATH_NOT_FOUND"
+	errCodeInvalidPath       = "INVALID_PATH"
+	errCodeInvalidOffset     = "INVALID_OFFSET"
+	errCodeReadOnlyFS        = "READ_ONLY_FILESYSTEM"
+	errCodeFileTooLarge      = "FILE_TOO_LARGE"
+	errCodeDeviceOutOfSpace  = "DEVICE_OUT_OF_SPACE"
+	errCodeUnknown           = "UNKNOWN"
+
+	// errnoEFBIG represents a file too large error.
+	// See `man errno` for more.
+	errnoEFBIG = 27
+	// errnoENOSPC represents a device out of space error.
+	// See `man errno` for more.
+	errnoENOSPC = 28
+	// errnoEROFS represents the syscall error number returned when an attempt
+	// to write to a read-only filesystem was made.
+	// See `man errno` for more.
+	errnoEROFS = 30
+)
+
+func storageErrorCode(err error) string {
+	switch actual := err.(type) {
+	case storagedriver.ErrUnsupportedMethod:
+		return errCodeUnsupportedMethod
+	case storagedriver.PathNotFoundError:
+		return errCodePathNotFound
+	case storagedriver.InvalidPathError:
+		return errCodeInvalidPath
+	case storagedriver.InvalidOffsetError:
+		return errCodeInvalidOffset
+	case storagedriver.Error:
+		perr, ok := actual.Enclosed.(*fs.PathError)
+		if !ok {
+			return errCodeUnknown
+		}
+		unwrapped := perr.Unwrap()
```

Comment:
I think this code block can be simplified to this;

```go
var perr *fs.PathError
		if !errors.As(actual.Enclosed, &perr) {
			return errCodeUnknown
		}

                if !errors.Is(perr), syscall.errno){
                     return errCodeUnknown
                }
		return syscallErrnoToErrorCode(perr.Unwrap().(syscall.Errno))
```
But, again, I haven't tested and may miss something :)

Reply:
I don't think we can use `errors.Is` with `syscall.Errno` because it's not an `error`, but a type alias to `uintptr`: https://pkg.go.dev/syscall#Errno.
If I try to call `errors.Is(perr, syscall.Errno)` I get the following compilation error:
```
pkg/dockerregistry/server/metrics/errorcodes.go:48:23: syscall.Errno (type) is not an expression
```

Reply:
ah, makes sense. Then it looks good to me.